File: frontend/src/types/request.ts
import dayjs from "dayjs";
// Types
import { MissingAttribute, ShiftWorkerOptionT } from "./constraint";
import { ColumnFilter, TableSort } from "./filter";

export enum RequestStatus {
  PENDING = "pending", // Waiting for manager review
  APPROVED = "approved", // Approved and must be fulfilled
  DENIED = "denied", // Denied and must not be fulfilled
  DEFERRED = "deferred", // Left to the algorithm to decide
}

export enum FulfillmentStatus {
  NOT_PROCESSED = "not_processed",
  FULFILLED = "fulfilled",
  UNFULFILLED = "unfulfilled",
}
export enum RequestType {
  WORK_DEMAND = "work_demand",
  LEAVE = "leave",
}

export type RequestT = {
  id: string;
  teamId: string;
  requestType: RequestType;
  workerId: string;
  startDate: dayjs.Dayjs;
  endDate: dayjs.Dayjs;
  shiftId: string | null;
  shiftOptions: ShiftWorkerOptionT[];
  negative: boolean;
  hard: boolean;
  status: RequestStatus;
  fulfillment: FulfillmentStatus;
  comment: string;
  createdAt: dayjs.Dayjs;
  active: boolean;
  shiftTargetIds: string[];
  missingAttributes: MissingAttribute[];
};

export const toRequestT = (data: any) => {
  const r: RequestT = {
    ...data,
    startDate: dayjs.unix(data.startDate).utc(),
    endDate: dayjs.unix(data.endDate).utc(),
    createdAt: dayjs.unix(data.createdAt).utc(),
  };
  return r;
};

export const fromRequestT = (data: RequestT) => {
  return {
    ...data,
    startDate: data.startDate.unix(),
    endDate: data.endDate.unix(),
    createdAt: data.createdAt.unix(),
  };
};

// Request Calendar View Settings
export type RequestCalendarViewSettingsT = {
  timeFrame: "week" | "month";
  periodStartDate: dayjs.Dayjs;
};

// Request View Settings (combines calendar and table state)
export type RequestViewSettingsT = {
  selectedTab: "table" | "calendar";
  filters: ColumnFilter[];
  sort: TableSort | null;
  timeFrame: "week" | "month";
  periodStartDate: dayjs.Dayjs;
};

File: backend/shared/src/shared/schemas/schemas/request.py
from dataclasses import asdict, dataclass
from datetime import date, datetime, time, timezone
from enum import Enum
from typing import Dict


class RequestStatus(Enum):
    PENDING = 0
    APPROVED = 1
    REJECTED = 2
    DISABLED = 3


@dataclass
class Request:
    id: str
    team_id: str
    worker_id: str
    start_date: date
    end_date: date
    shift_id: str
    negative: bool
    hard: bool
    status: RequestStatus

    def to_dict(self) -> Dict:
        out = asdict(self)
        out["start_date"] = datetime.combine(
            self.start_date, time.min, tzinfo=timezone.utc
        ).timestamp()
        out["end_date"] = datetime.combine(
            self.end_date, time.min, tzinfo=timezone.utc
        ).timestamp()
        out["status"] = self.status.value
        return out

    @classmethod
    def from_dict(cls, data: Dict) -> "Request":
        return cls(
            id=data["id"],
            team_id=data["team_id"],
            worker_id=data["worker_id"],
            start_date=datetime.fromtimestamp(
                data["start_date"], tz=timezone.utc
            ).date(),
            end_date=datetime.fromtimestamp(
                data["end_date"], tz=timezone.utc
            ).date(),
            shift_id=data["shift_id"],
            negative=data["negative"],
            hard=data["hard"],
            status=RequestStatus(data["status"]),
        )


@dataclass
class RequestAugmented(Request):
    active: bool

    def to_dict(self) -> Dict:
        out = super().to_dict()
        out.update({"active": self.active})
        return out

    @classmethod
    def from_dict(cls, data: Dict) -> "RequestAugmented":
        return cls(
            id=data["id"],
            team_id=data["team_id"],
            worker_id=data["worker_id"],
            start_date=datetime.fromtimestamp(
                data["start_date"], tz=timezone.utc
            ).date(),
            end_date=datetime.fromtimestamp(
                data["end_date"], tz=timezone.utc
            ).date(),
            shift_id=data["shift_id"],
            negative=data["negative"],
            hard=data["hard"],
            status=RequestStatus(data["status"]),
            active=data["active"],
        )
